server/src/main/java/org/elasticsearch/index/mapper/DocumentParserContext.java

```diff
@@ -287,6 +301,59 @@ public final Collection<IgnoredSourceFieldMapper.NameValue> getIgnoredFieldValue
         return Collections.unmodifiableCollection(ignoredFieldValues);
     }
 
+    /**
+     * Adds an ignored field from the parser context, capturing an object or an array.
+     *
+     * In case of nested arrays, i.e. capturing an array within an array, elements tracked as ignored fields may interfere with
+     * the rest, as ignored source contents take precedence over regular field contents with the same leaf name. To prevent
+     * missing array elements from synthetic source, all array elements get recorded in ignored source. Otherwise, just the value in
+     * current parsing context gets captured.
+     *
+     * In both cases, a new parser sub-context gets created from the current {@link DocumentParserContext} and returned, indicating
+     * that the source for the sub-context has been captured, to avoid double-storing parts of its contents to ignored source.
+     */
+    public final DocumentParserContext addIgnoredFieldFromContext(IgnoredSourceFieldMapper.NameValue ignoredFieldWithNoSource)
+        throws IOException {
+        if (canAddIgnoredField()) {
+            if (parentArrayField != null
+                && parent != null
+                && parentArrayField.equals(parent.fullPath())
```

Comment:
Hm this now actually works for one level, as the parentArrayField doesn't get updated.. Lemme think about it.

Reply:
Updated to track the name of the last array. I'll add some yaml tests to verify this works as intended.